FLINT 3.3.1 FTBFS on ubuntu-bionic-toolchain-gcc_9-i386

```
  [flint-3.3.1]   [spkg-install] In file included from /sage/local/var/tmp/sage/build/flint-3.3.1/src/src/longlong.h:50,
  [flint-3.3.1]   [spkg-install]                  from /sage/local/var/tmp/sage/build/flint-3.3.1/src/src/mpn_extras.h:25,
  [flint-3.3.1]   [spkg-install]                  from /sage/local/var/tmp/sage/build/flint-3.3.1/src/src/nfloat/nfixed.c:13:
  [flint-3.3.1]   [spkg-install] /sage/local/var/tmp/sage/build/flint-3.3.1/src/src/nfloat/nfixed.c: In function '_nfixed_vec_sub':
  [flint-3.3.1]   [spkg-install] /sage/local/var/tmp/sage/build/flint-3.3.1/src/src/longlong_asm_gcc.h:84:3: error: 'asm' operand has impossible constraints
  [flint-3.3.1]   [spkg-install]    84 |   __asm__(_ASM_ADD " %17,%" _ASM_PRE "5\n" \
  [flint-3.3.1]   [spkg-install]       |   ^~~~~~~
  [flint-3.3.1]   [spkg-install] /sage/local/var/tmp/sage/build/flint-3.3.1/src/src/mpn_extras.h:198:27: note: in expansion of macro 'add_ssssssaaaaaaaaaaaa'
  [flint-3.3.1]   [spkg-install]   198 | #define NN_ADD_6(r, u, v) add_ssssssaaaaaaaaaaaa((r)[5], (r)[4], (r)[3], (r)[2], (r)[1], (r)[0], (u)[5], (u)[4], (u)[3], (u)[2], (u)[1], (u)[0], (v)[5], (v)[4], (v)[3], (v)[2], (v)[1], (v)[0])
  [flint-3.3.1]   [spkg-install]       |                           ^~~~~~~~~~~~~~~~~~~~~~
  [flint-3.3.1]   [spkg-install] /sage/local/var/tmp/sage/build/flint-3.3.1/src/src/nfloat/nfixed.c:91:9: note: in expansion of macro 'NN_ADD_6'
  [flint-3.3.1]   [spkg-install]    91 |         NN_ADD_ ## n(res + 1, a + 1, b + 1); \
  [flint-3.3.1]   [spkg-install]       |         ^~~~~~~
  [flint-3.3.1]   [spkg-install] /sage/local/var/tmp/sage/build/flint-3.3.1/src/src/nfloat/nfixed.c:111:1: note: in expansion of macro 'DEF_NFIXED_SUB'
  [flint-3.3.1]   [spkg-install]   111 | DEF_NFIXED_SUB(6)
  [flint-3.3.1]   [spkg-install]       | ^~~~~~~~~~~~~~
  [flint-3.3.1]   [spkg-install] /sage/local/var/tmp/sage/build/flint-3.3.1/src/src/longlong_asm_gcc.h:84:3: error: 'asm' operand has impossible constraints
  [flint-3.3.1]   [spkg-install]    84 |   __asm__(_ASM_ADD " %17,%" _ASM_PRE "5\n" \
  [flint-3.3.1]   [spkg-install]       |   ^~~~~~~
  [flint-3.3.1]   [spkg-install] /sage/local/var/tmp/sage/build/flint-3.3.1/src/src/longlong.h:204:5: note: in expansion of macro 'add_ssssssaaaaaaaaaaaa'
  [flint-3.3.1]   [spkg-install]   204 |     add_ssssssaaaaaaaaaaaa(__t6, s4, s3, s2, s1, s0, (ulong) 0, a4, a3, a2, a1, a0, (ulong) 0, b4, b3, b2, b1, b0); \
  [flint-3.3.1]   [spkg-install]       |     ^~~~~~~~~~~~~~~~~~~~~~
```
https://github.com/passagemath/passagemath/actions/runs/16460404700/job/46526402882#step:11:7037

Possibly related:  We got the following on the armhf build of the Debian package of flint 3.3.1:

```m2
  CC  fmpz_mod/ctx.c
In file included from /build/reproducible-path/flint-3.3.1/src/longlong.h:50,
                 from /build/reproducible-path/flint-3.3.1/src/fmpz_poly/taylor_shift_horner.c:13:
/build/reproducible-path/flint-3.3.1/src/fmpz_poly/taylor_shift_horner.c: In function ‘_fmpz_poly_taylor_shift_horner’:
/build/reproducible-path/flint-3.3.1/src/longlong_asm_gcc.h:266:3: error: ‘asm’ operand has impossible constraints or there are not enough registers
  266 |   __asm__("adds %2,%5,%8\n" \
      |   ^~~~~~~
/build/reproducible-path/flint-3.3.1/src/fmpz_poly/taylor_shift_horner.c:143:21: note: in expansion of macro ‘add_sssaaaaaa’
  143 |                     add_sssaaaaaa(t[3 * j + 2], t[3 * j + 1], t[3 * j],
      |                     ^~~~~~~~~~~~~
  CC  fmpz_mod/discrete_log_pohlig_hellman.c
```

Full log: https://buildd.debian.org/status/fetch.php?pkg=flint&arch=armhf&ver=3.3.1-1~exp1&stamp=1756320704&file=log

--- a/src/longlong.h
+++ b/src/longlong.h
@@ -128,7 +128,9 @@ flint_bitcnt_t FLINT_BIT_COUNT(ulong x)
     (s0) = (a0) - (b0); \
     (s1) = (a1) - (b1) - ((ulong) (s0) > __t0); \
   } while (0)
+#endif
 
+#if !defined(add_sssaaaaaa)
 # define add_sssaaaaaa(s2, s1, s0, a2, a1, a0, b2, b1, b0) \
   do { \
     ulong __t1, __t2; \
@@ -162,7 +164,9 @@ flint_bitcnt_t FLINT_BIT_COUNT(ulong x)
     sub_ddmmss(__t4, s2, (ulong) 0, a2, (ulong) 0, b2); \
     sub_ddmmss(s3, s2, (a3) - (b3), s2, -__t4, -__t3); \
   } while (0)
+#endif
 
+#if !defined(add_sssssaaaaaaaaaa)
 # define add_sssssaaaaaaaaaa(s4, s3, s2, s1, s0, a4, a3, a2, a1, a0, b4, b3, b2, b1, b0) \
   do { \
     ulong __t4 = 0; \
@@ -178,7 +182,9 @@ flint_bitcnt_t FLINT_BIT_COUNT(ulong x)
     sub_ddmmss(__t6, s3, (ulong) 0, a3, (ulong) 0, b3); \
     sub_ddmmss(s4, s3, (a4) - (b4), s3, -__t6, -__t5); \
   } while (0)
+#endif
 
+#if !defined(add_ssssssaaaaaaaaaaaa)
 # define add_ssssssaaaaaaaaaaaa(s5, s4, s3, s2, s1, s0, a5, a4, a3, a2, a1, a0, b5, b4, b3, b2, b1, b0) \
   do { \
     ulong __t5 = 0; \
@@ -196,7 +202,6 @@ flint_bitcnt_t FLINT_BIT_COUNT(ulong x)
   } while (0)
 #endif
 
-/* extra wide variants might not have inline asm if there are not enough registers */
 #if !defined(add_sssssssaaaaaaaaaaaaaa)
 # define add_sssssssaaaaaaaaaaaaaa(s6, s5, s4, s3, s2, s1, s0, a6, a5, a4, a3, a2, a1, a0, b6, b5, b4, b3, b2, b1, b0) \
   do { \
@@ -206,20 +211,22 @@ flint_bitcnt_t FLINT_BIT_COUNT(ulong x)
     add_ssaaaa(s6, s5, s6, s5, (ulong) 0, __t6); \
   } while (0)
 
-# define add_ssssssssaaaaaaaaaaaaaaaa(s7, s6, s5, s4, s3, s2, s1, s0, a7, a6, a5, a4, a3, a2, a1, a0, b7, b6, b5, b4, b3, b2, b1, b0) \
-  do { \
-    ulong __t7 = 0; \
-    add_sssssssaaaaaaaaaaaaaa(__t7, s5, s4, s3, s2, s1, s0, (ulong) 0, a5, a4, a3, a2, a1, a0, (ulong) 0, b5, b4, b3, b2, b1, b0); \
-    add_ssaaaa(s7, s6, a7, a6, b7, b6); \
-    add_ssaaaa(s7, s6, s7, s6, (ulong) 0, __t7); \
-  } while (0)
-
 # define sub_dddddddmmmmmmmsssssss(s6, s5, s4, s3, s2, s1, s0, a6, a5, a4, a3, a2, a1, a0, b6, b5, b4, b3, b2, b1, b0) \
   do { \
     ulong __t9, __t10; \
     sub_ddddddmmmmmmssssss(__t9, s4, s3, s2, s1, s0, (ulong) 0, a4, a3, a2, a1, a0, (ulong) 0, b4, b3, b2, b1, b0); \
     sub_ddmmss(__t10, s5, (ulong) 0, a5, (ulong) 0, b5); \
     sub_ddmmss(s6, s5, (a6) - (b6), s5, -__t10, -__t9); \
+  } while (0)
+#endif
+
+#if !defined(add_ssssssssaaaaaaaaaaaaaaaa)
+# define add_ssssssssaaaaaaaaaaaaaaaa(s7, s6, s5, s4, s3, s2, s1, s0, a7, a6, a5, a4, a3, a2, a1, a0, b7, b6, b5, b4, b3, b2, b1, b0) \
+  do { \
+    ulong __t7 = 0; \
+    add_sssssssaaaaaaaaaaaaaa(__t7, s5, s4, s3, s2, s1, s0, (ulong) 0, a5, a4, a3, a2, a1, a0, (ulong) 0, b5, b4, b3, b2, b1, b0); \
+    add_ssaaaa(s7, s6, a7, a6, b7, b6); \
+    add_ssaaaa(s7, s6, s7, s6, (ulong) 0, __t7); \
   } while (0)
 
 # define sub_ddddddddmmmmmmmmssssssss(s7, s6, s5, s4, s3, s2, s1, s0, a7, a6, a5, a4, a3, a2, a1, a0, b7, b6, b5, b4, b3, b2, b1, b0) \
@@ -230,7 +237,6 @@ flint_bitcnt_t FLINT_BIT_COUNT(ulong x)
     sub_ddmmss(s7, s6, (a7) - (b7), s6, -__t12, -__t11); \
   } while (0)
 #endif
-
 
 #if !defined(MPN_INCR_U)
 # if FLINT_WANT_ASSERT

--- a/src/longlong_asm_gcc.h
+++ b/src/longlong_asm_gcc.h
@@ -80,21 +80,6 @@
       "3" ((ulong)(a1)), _ASM_RME ((ulong)(b1)), \
       "4" ((ulong)(a0)), _ASM_RME ((ulong)(b0)))
 
-# define add_ssssssaaaaaaaaaaaa(s5, s4, s3, s2, s1, s0, a5, a4, a3, a2, a1, a0, b5, b4, b3, b2, b1, b0) \
-  __asm__(_ASM_ADD " %17,%" _ASM_PRE "5\n" \
-     "\t" _ASM_ADC " %15,%" _ASM_PRE "4\n" \
-     "\t" _ASM_ADC " %13,%" _ASM_PRE "3\n" \
-     "\t" _ASM_ADC " %11,%" _ASM_PRE "2\n" \
-     "\t" _ASM_ADC " %9,%" _ASM_PRE "1\n" \
-     "\t" _ASM_ADC " %7,%" _ASM_PRE "0" \
-    : "=r" (s5), "=&r" (s4), "=&r" (s3), "=&r" (s2), "=&r" (s1), "=&r" (s0) \
-    : "0" ((ulong)(a5)), _ASM_RME ((ulong)(b5)), \
-      "1" ((ulong)(a4)), _ASM_RME ((ulong)(b4)), \
-      "2" ((ulong)(a3)), _ASM_RME ((ulong)(b3)), \
-      "3" ((ulong)(a2)), _ASM_RME ((ulong)(b2)), \
-      "4" ((ulong)(a1)), _ASM_RME ((ulong)(b1)), \
-      "5" ((ulong)(a0)), _ASM_RME ((ulong)(b0)))
-
 # define sub_ddmmss(d1, d0, m1, m0, s1, s0) \
   __asm__(_ASM_SUB " %5,%" _ASM_PRE "1\n" \
      "\t" _ASM_SBB " %3,%" _ASM_PRE "0" \
@@ -135,23 +120,23 @@
       "3" ((ulong)(m1)), _ASM_RME ((ulong)(s1)), \
       "4" ((ulong)(m0)), _ASM_RME ((ulong)(s0)))
 
-# define sub_ddddddmmmmmmssssss(d5, d4, d3, d2, d1, d0, m5, m4, m3, m2, m1, m0, s5, s4, s3, s2, s1, s0) \
-  __asm__(_ASM_SUB " %17,%" _ASM_PRE "5\n" \
-     "\t" _ASM_SBB " %15,%" _ASM_PRE "4\n" \
-     "\t" _ASM_SBB " %13,%" _ASM_PRE "3\n" \
-     "\t" _ASM_SBB " %11,%" _ASM_PRE "2\n" \
-     "\t" _ASM_SBB " %9,%" _ASM_PRE "1\n" \
-     "\t" _ASM_SBB " %7,%" _ASM_PRE "0" \
-    : "=r" (d5), "=&r" (d4), "=&r" (d3), "=&r" (d2), "=&r" (d1), "=&r" (d0) \
-    : "0" ((ulong)(m5)), _ASM_RME ((ulong)(s5)), \
-      "1" ((ulong)(m4)), _ASM_RME ((ulong)(s4)), \
-      "2" ((ulong)(m3)), _ASM_RME ((ulong)(s3)), \
-      "3" ((ulong)(m2)), _ASM_RME ((ulong)(s2)), \
-      "4" ((ulong)(m1)), _ASM_RME ((ulong)(s1)), \
-      "5" ((ulong)(m0)), _ASM_RME ((ulong)(s0)))
-
 /* x86 does not have enough registers */
 # if FLINT_BITS == 64 && defined (__amd64__)
+
+# define add_ssssssaaaaaaaaaaaa(s5, s4, s3, s2, s1, s0, a5, a4, a3, a2, a1, a0, b5, b4, b3, b2, b1, b0) \
+  __asm__(_ASM_ADD " %17,%" _ASM_PRE "5\n" \
+     "\t" _ASM_ADC " %15,%" _ASM_PRE "4\n" \
+     "\t" _ASM_ADC " %13,%" _ASM_PRE "3\n" \
+     "\t" _ASM_ADC " %11,%" _ASM_PRE "2\n" \
+     "\t" _ASM_ADC " %9,%" _ASM_PRE "1\n" \
+     "\t" _ASM_ADC " %7,%" _ASM_PRE "0" \
+    : "=r" (s5), "=&r" (s4), "=&r" (s3), "=&r" (s2), "=&r" (s1), "=&r" (s0) \
+    : "0" ((ulong)(a5)), _ASM_RME ((ulong)(b5)), \
+      "1" ((ulong)(a4)), _ASM_RME ((ulong)(b4)), \
+      "2" ((ulong)(a3)), _ASM_RME ((ulong)(b3)), \
+      "3" ((ulong)(a2)), _ASM_RME ((ulong)(b2)), \
+      "4" ((ulong)(a1)), _ASM_RME ((ulong)(b1)), \
+      "5" ((ulong)(a0)), _ASM_RME ((ulong)(b0)))
 
 # define add_sssssssaaaaaaaaaaaaaa(s6, s5, s4, s3, s2, s1, s0, a6, a5, a4, a3, a2, a1, a0, b6, b5, b4, b3, b2, b1, b0) \
   __asm__(_ASM_ADD " %20,%" _ASM_PRE "6\n" \
@@ -188,6 +173,21 @@
       "5" ((ulong)(a2)), _ASM_RME ((ulong)(b2)), \
       "6" ((ulong)(a1)), _ASM_RME ((ulong)(b1)), \
       "7" ((ulong)(a0)), _ASM_RME ((ulong)(b0)))
+
+# define sub_ddddddmmmmmmssssss(d5, d4, d3, d2, d1, d0, m5, m4, m3, m2, m1, m0, s5, s4, s3, s2, s1, s0) \
+  __asm__(_ASM_SUB " %17,%" _ASM_PRE "5\n" \
+     "\t" _ASM_SBB " %15,%" _ASM_PRE "4\n" \
+     "\t" _ASM_SBB " %13,%" _ASM_PRE "3\n" \
+     "\t" _ASM_SBB " %11,%" _ASM_PRE "2\n" \
+     "\t" _ASM_SBB " %9,%" _ASM_PRE "1\n" \
+     "\t" _ASM_SBB " %7,%" _ASM_PRE "0" \
+    : "=r" (d5), "=&r" (d4), "=&r" (d3), "=&r" (d2), "=&r" (d1), "=&r" (d0) \
+    : "0" ((ulong)(m5)), _ASM_RME ((ulong)(s5)), \
+      "1" ((ulong)(m4)), _ASM_RME ((ulong)(s4)), \
+      "2" ((ulong)(m3)), _ASM_RME ((ulong)(s3)), \
+      "3" ((ulong)(m2)), _ASM_RME ((ulong)(s2)), \
+      "4" ((ulong)(m1)), _ASM_RME ((ulong)(s1)), \
+      "5" ((ulong)(m0)), _ASM_RME ((ulong)(s0)))
 
 # define sub_dddddddmmmmmmmsssssss(d6, d5, d4, d3, d2, d1, d0, m6, m5, m4, m3, m2, m1, m0, s6, s5, s4, s3, s2, s1, s0) \
   __asm__(_ASM_SUB " %20,%" _ASM_PRE "6\n" \

